// lib/dna-to-cells.js
const loadDNA = require('organic-dna-loader')
const cellsinfo = require('organic-dna-cells-info')
const {
  Cell
} = require('lib/chemicals')
const path = require('path')

module.exports = function (cwd, existingCells = []) {
  return new Promise((resolve, reject) => {
    loadDNA({
      dnaSourcePaths: [
        path.join(cwd, 'dna')
      ]
    }, (err, dna) => {
      if (err) {
        console.error(err)
        return reject(err)
      }
      if (!dna.cells) {
        console.log(dna)
        reject(new Error('failed to locate cells in ' + cwd + '/dna'))
      }
      let found_cells = cellsinfo(dna.cells)
      let updated_cells = found_cells.map((cellInfo) => {
        let scripts = []
        if (cellInfo.dna.cwd) {
          try {
            scripts = require(path.join(cwd, cellInfo.dna.cwd, 'package.json')).scripts
          } catch (e) {
            // ignore
          }
        }
        let existingCell = {}
        for (let i = 0; i < existingCells.length; i ++) {
          if (existingCells[i].name === cellInfo.name) {
            existingCell = existingCells[i]
          }
        }
        return Object.assign(existingCell, Cell.create({
          name: cellInfo.name,
          groups: cellInfo.groups,
          cwd: cellInfo.dna.cwd,
          selected: false,
          focused: false,
          commandRunning: false,
          port: dna['cell-ports'] ? dna['cell-ports'][cellInfo.name] : false,
          mountPoint: dna['cell-mountpoints'] ? dna['cell-mountpoints'][cellInfo.name] : false,
          scripts: scripts
        }))
      })
      resolve(updated_cells)
    })
  })
}
